crates/telio-firewall/src/firewall.rs

```diff
@@ -1678,6 +1715,59 @@ pub mod tests {
         }
     }
 
+    #[test]
+    fn firewall_icmp_cache_error_response() {
+        struct TestInput(
+            &'static str,
+            &'static str,
+            GenericIcmpType,
+            GenericIcmpType,
+            MakeIcmpWithBody,
+        );
+        for TestInput(src, dst, request_type, error_type, make_icmp) in [
+            TestInput(
+                "8.8.8.8",
+                "127.0.0.1",
+                IcmpTypes::EchoRequest.into(),
+                IcmpTypes::DestinationUnreachable.into(),
+                &make_icmp4_with_body,
+            ),
+            TestInput(
+                "2001:4860:4860::8888",
+                "::1",
+                Icmpv6Types::EchoRequest.into(),
+                Icmpv6Types::DestinationUnreachable.into(),
+                &make_icmp6_with_body,
+            ),
+        ] {
+            let fw = StatefullFirewall::new();
+
+            let actual_request = make_icmp(src, dst, request_type, &[1, 0, 1, 0]);
+            let other_request = make_icmp(src, dst, request_type, &[2, 0, 3, 0]);
+
+            let mut actual_response_body = vec![0; 4];
+            actual_response_body.extend_from_slice(&actual_request);
+            let actual_response = make_icmp(dst, src, error_type, &actual_response_body);
+
+            let mut other_response_body = vec![0; 4];
+            other_response_body.extend_from_slice(&other_request);
+            let other_response = make_icmp(dst, src, error_type, &other_response_body);
+
+            assert_eq!(
+                fw.process_outbound_packet(&make_peer(), &actual_request),
+                true
+            );
+            assert_eq!(
+                fw.process_inbound_packet(&make_peer(), &other_response),
+                false
+            );
+            assert_eq!(
+                fw.process_inbound_packet(&make_peer(), &actual_response),
+                true
+            );
+        }
+    }
+
     #[rustfmt::skip]
```

Comment:
Keeping in mind https://github.com/NordSecurity/libtelio/pull/159/files#r1360652683, maybe it would make sense to add two tests where ICMP in its own payload contains UDP packet in one case and then TCP packet in another case?